### DISCUSSION TITLE
calculate luminosity

### DISCUSSION THREAD

**Discussion by exekutive** [ImageMagick/ImageMagick#7395] [Help] [answered]
### ImageMagick version

7.1.1-31 CLI

### Operating system, version and so on

darwin AArch64

### Description

Given a color value as output by IM , I need to store it, plus determine if it's closer to black or white.
What's the most efficient way to do this? Does IM have a function?

value comes in different forms. Eg...
```
(11795,11795,11795) #2E132E132E13 gray(17.998%)
(245,243,239) #F5F3EF srgb(245,243,239)
(255,250,240) #FFFAF0 FloralWhite
(22798,20542,17484) #590E503E444C srgb(34.7875%,31.3451%,26.6789%)
(219,219,219) #DBDBDB gray(219)
```

My idea is to parse out the last parameter by piping to awk:
`col=$( magick ... | awk '{ printf "%s", $3 }' )`

then use IM threshold to output black or white
`magick canvas:"$col" -threshold  50% -format '%[pixel:s]' info:-`

**Comment by fmw42** [comment]
Once you have your color, you can do the following.   It sets the intensity of the color according to any one of several measures. See https://imagemagick.org/script/command-line-options.php#intensity.  Then compares the value (in range 0 to 1) to 0.5.  If larger, it sets the output to 1, if smaller it sets the output to 0.  Then there is a an IF test to convert the 0 or 1 to a text message. 

I do not know if your version of IM supports -intensity XXX, but if not just replace that with -colorspace gray.

```
color="rgb(200,100,50)"
test=`convert xc:"$color" -alpha off -intensity Rec709Luma -format "%[fx:u>0.5?1:0]" info:`
if [ $test -eq 1 ]; then
echo "image closer to white"
else
echo "image closer to white"
fi

```

**Comment by exekutive** [comment]
Neat. I didn't know IM could do logic.
Looks like 7.1.1 supports the intensity operator. Is this method superior to using threshold?
I couldn't find in the docs whether threshold uses any gamma correction.
Thank you.

```
magick -list intensity 
Average
Brightness
Lightness
Mean
MS
Rec601Luma
Rec601Luminance
Rec709Luma
Rec709Luminance
RMS
```

**Reply by fmw42** [reply]
These are just options for converting your color image to grayscale before the mathematics operation in %[fx...]. Each one weights the color channels (RGB) differently. We just call that function -intensity.

I think thresholding adds an extra layer of processing that you do not need. You would still have to compute the mean (i.e.average) to compare to mid-gray. I note that thresholding the color image does a conversion to intensity before it applies the threshold. So it is thresholding on the intensity image already.  The intensity conversion is using Rec601Luma or Rec709Luma depending upon IM 6 or IM 7, respectively, I believe. So by thresholding you would be doing one conversion intensity conversion for the threshold and an average to get a value to compare to mid-gray.

**Comment by snibgo** [comment, marked as answer]
fmw42 has suggested:
```
test=`convert xc:"$color" -alpha off -intensity Rec709Luma -format "%[fx:u>0.5?1:0]" info:`
```
I suggest use `magick` instead of `convert`.

But there is also an error. `-intensity Rec709Luma` is a setting. It doesn't change the image. It merely changes the behaviour of some operations, but the command has no operations affected by `-intensity`.

`-format "%[fx:u>0.5?1:0]"` will return `0` or `1`, according to whether the *first* channel is less than or more than 0.5 (on a scale of 0 to 1). For an RGB image, the first channel is the red channel.

We can demonstrate the error:
```
magick xc:sRGB(10%,20%,30%) -intensity Rec709Luma -format "%[fx:u]" info:
0.1
```
The required intensity should be somewhere between 0.1 and 0.3. But the command simply returns the red channel, which is 0.1.

We can get the required intensity like this:
```
magick xc:sRGB(10%,20%,30%) -intensity Rec709Luma -format "%[fx:intensity]" info:
0.185953
```
In this command, the result of `%[fx:intensity]` is determined by the `-intensity` setting. For example, using a different `-intensity` setting:
```
magick xc:sRGB(10%,20%,30%) -intensity Rec601Luma -format "%[fx:intensity]" info:
0.181551
```

**Comment by fmw42** [comment]
Sorry, my command should have been -grayscale rather than -intensity.  My mistake as snibgo has pointed out. If on IM 7, use magick rather than convert

`convert xc:"$color" -alpha off -grayscale Rec709Luma -format "%[fx:u>0.5?1:0]" info:`

**Comment by exekutive** [comment]
Thank you for the clarification.

Actually, the command

`magick xc:sRGB(10%,20%,30%) -intensity Rec709Luma -format "%[fx:intensity]" info:`

returns
`zsh: unknown file attribute: 1`

This works better:

`magick xc:"sRGB(10%,20%,30%)" -intensity Rec709Luma -format "%[fx:intensity]" info:`
